### ISSUE TITLE
[macOS][bug] verticalSync problem since macOS13 (GLWF-inherited issue)

### ISSUE THREAD

**Issue by artificiel** [openframeworks/openFrameworks#7760]
macOS13 introduces jitter in the timing of update/draw cycle of GLFW apps https://github.com/glfw/glfw/issues/2249. it has (at least) 2 concrete negative effects in OF as demontrated in https://github.com/openframeworks/openFrameworks/issues/7752#issuecomment-1817035425

- sometimes the jitter pushes the `draw()` later than the actual frame cycle, stuttering/dropping a frame

 - `ofGetElapsedTimef()` is not reliable to control phase-based animations requiring isochronous steps

until it is fixed within GLFW ([seems hard](https://github.com/RustyMoyher/glfw_MacVsyncFix)) — or at OF-level by working around the Timer ([as inspiration see the google deepmind workaround](https://github.com/google-deepmind/mujoco/commit/2f0fb1e4ef81ae4eb281e09fb67c6b36f27dc427)) , or by implementing a metal backend, or other — `ofSetVerticalSync(false)` and using `ofSetFrameRate(nnn)` is recommended. using [macOS14 game mode in fullscreen](https://eclecticlight.co/2023/10/18/how-game-mode-manages-cpu-and-gpu/)  also helps as seen in https://github.com/openframeworks/openFrameworks/issues/7752#issuecomment-1817035425.

the uneven timing issue can be worked around by using something like [`ofGetExpectedPaintTimef()`](https://github.com/openframeworks/openFrameworks/issues/7758), but that will not prevent dropped frames.

this issue will be updated as the GLFW and OF evolution affects the behaviour.

**Comment by danomatika** [comment]
Does this issue happen with both `ofSetVerticalSync(true)` and `ofSetFrameRate(60)`? I don't *quite* remember why, but I feel like there was an issue with sync only and I added setting the frame rate as well.

**Comment by artificiel** [comment]
yes, adding the framerate is better top the point of things being mostly unnoticeable, if the CPU load is low and if you don't relay of ofGetElapsedTimef()

timing with vsync, frameRate(60), OF fullscreen:
<img width="1374" alt="image" src="https://github.com/openframeworks/openFrameworks/assets/2172370/d10d82de-eba6-4a33-987e-74ab2a67ff35">

timing without vsync,  frameRate(60), OF fullscreen:
<img width="1346" alt="image" src="https://github.com/openframeworks/openFrameworks/assets/2172370/387da31b-cff7-47b0-bad3-703f2bd01a04">

and to give good measure of what is achievable, timing native fullscreen with game mode and no vsync:
<img width="1356" alt="image" src="https://github.com/openframeworks/openFrameworks/assets/2172370/f4f1dc8c-e929-448a-9ead-1e3178356bb3">

(more can be found [here](https://github.com/openframeworks/openFrameworks/issues/7752))

**Comment by dimitre** [comment]
I think ofSetFrameRate should be redesigned out of ofTimer.
Or maybe an alternative ofTimer using std::chrono exclusively.
I've made some experiments in measuring here:
- https://github.com/openframeworks/openFrameworks/issues/7866

and some insights from Glenn Fiedler about fps https://gafferongames.com/post/fix_your_timestep/

**Comment by dimitre** [comment]
Framerate drift is merged and it is much more precise now. 
VSync is doable now outside GLFW if we use a similar solution to this one:
https://github.com/google-deepmind/mujoco/commit/2f0fb1e4ef81ae4eb281e09fb67c6b36f27dc427
It seems ok because OF project is already including CoreVideo framework

**Comment by danomatika** [comment]
Thanks for the work on this, all!

**Comment by dimitre** [comment]
@artificiel do you still have this tests to re-run ?

**Comment by artificiel** [comment]
@dimitre it needed some mods to the core to get hooks in the right place; will look into that and report.

also if you're into time stuff, exploring around this idea https://github.com/openframeworks/openFrameworks/issues/7758 is still meaningful for dt-accurate animations. (the pseudocode returns float but of course it should be proper time point)
